### ISSUE TITLE
endless while loop in estimateLibrarySize() in DuplicationMetrics.java

### ISSUE THREAD

**Issue by zhangs3** [broadinstitute/picard#1146]
### Affected tool(s)

picard MarkDuplicates

### Affected version(s)

Latest public release version: 2.18.0

### Description 

The while loop in the following codes (in the function estimateLibrarySize() in DuplicationMetrics.java) never ends for some initial values of M, c and n with which the function f() will never return a negative value.

    while( f(M*c, c, n) >= 0 ) M *= 10.0;
     .
     .
     .
    private static double f(double x, double c, double n) {
            return c/x - 1 + Math.exp(-n/x);
    }

#### Steps to reproduce

Here is a short Java program to show the problem. The initial values of M, c and n are from real data.

    public class Loop
    {
        public static void main(String[] args)
        {
            // bug!!! f() never returns negative for initial values of
            double M = 100.0;
            double c = 357087881;
            double n = 357087883;
    
            while( f(M*c, c, n) >= 0 ) M *= 10.0; 
        }
    
        private static double f(double x, double c, double n) {
            System.err.println(c/x - 1 + Math.exp(-n/x));
            return c/x - 1 + Math.exp(-n/x);
        }
    }

#### Expected behavior

The while loop should exit for certain value of M.

#### Actual behavior

For certain values of M, c and n, such as, 100.0, 357087881 and 357087883 respectively, the function f() never returns a negative value.

#### Suggested correction

    while( f(M*c, c, n) > 0 ) M *= 10.0; 
    M *= 10.0;  // if necessary

**Comment by yfarjoun** [comment]
Thanks for this interesting edge case. 

I took your example and made a test out of it, and also fixed it in line with your suggestion.
(PR here #1147)
Thanks again!

**Comment by zhangs3** [comment]
Hello,

Here’s something else that I’m wondering if it’s normal or not.

I’ve been using picard for a couple of years with WES data, but never hit the bug until recently with WGS data.  After I found out the specific input values to the f() function that never came back as negative, I tried to see what values of c and n the program produced with my WES data and found that the function estimateLibrarySize() was never called because the for loop in the function finalizeAndWriteMetrics() was never entered and the metrics file produced contained no histogram data.  So my question is, is this normal? That is, is it expected that for some data the histogram will not be produced?  In our analysis pipelines, the metrics files are not used at all (that’s why I never checked the metrics files before).

Thanks.
